### PR TITLE
Phase2-hgx110 HGCal TDR geometry for SIM step

### PR DESCRIPTION
**Pull request by bsunanda**
Added functionality to get numbering in SIM step for new HGCal geometry definition
New sensitive detector class
Tested detid to space point and vice versa

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Review by kpedro88 — COMMENTED** [review]
Only minor comments on the geometry part, sim part needs some elaboration.

File: SimG4CMS/Calo/interface/HGCalSD.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#ifndef SimG4CMS_HGCalSD_h
+#define SimG4CMS_HGCalSD_h
+///////////////////////////////////////////////////////////////////////////////
+// File: HGCalSD.h
+// Description: Stores hits of the High Granularity Calorimeter (HGC) in the
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply SimG4CMS/Calo/interface/HGCalSD.h]: This is the new one - HGCSD is preserved for backward compatibility

File: SimG4CMS/Calo/src/HGCalSD.cc [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+      aStep->GetPreStepPoint()->GetPhysicalVolume()->GetLogicalVolume();
+    edm::LogVerbatim("HGCSim") << "HGCalSD: Hit from standard path from "
+			       << lv->GetName() << " for Track " 
+			       << aStep->GetTrack()->GetTrackID() << " ("
+			       << aStep->GetTrack()->GetDefinition()->GetParticleName() 
+			       << ":" << notaMuon << ") R = " << r << " Z = "
+			       << z << " slope = " << r/z << ":" << slopeMin_;
+#endif
+    // Apply fiducial cuts
+    if (r/z >= slopeMin_) {
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply SimG4CMS/Calo/src/HGCalSD.cc]: For any thing coming to HGCalSD z cannot be 0

File: SimG4CMS/Calo/src/HGCalSD.cc [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  if (hdc.isValid()) {
+    const HGCalDDDConstants* hgcons = hdc.product();
+    m_mode           = hgcons->geomMode();
+    slopeMin_        = hgcons->minSlope();
+    levelT1_         = hgcons->levelTop(0);
+    levelT2_         = hgcons->levelTop(1);
+    isScint_         = (m_mode == HGCalGeometryMode::Trapezoid);
+    double waferSize = hgcons->waferSize(false);
+    double mouseBite = hgcons->mouseBite(false);
+    mouseBiteCut_    = waferSize*tan(30.0*CLHEP::deg) - mouseBite;
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply SimG4CMS/Calo/src/HGCalSD.cc]: Once in a production

File: SimG4CMS/Calo/src/HGCalNumberingScheme.cc [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  edm::LogVerbatim("HGCSim") << "Deleting HGCalNumberingScheme";
+#endif
+}
+
+uint32_t HGCalNumberingScheme::getUnitID(int layer, int module, int cell,
+					 int iz, const G4ThreeVector &pos,
+					 double& wt) {
+  // module is the copy number of the wafer as placed in the layer
+  uint32_t index(0);
+  wt = 1.0;
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply SimG4CMS/Calo/src/HGCalNumberingScheme.cc]: It is not reset - wt is a return argument

File: SimG4CMS/Calo/interface/HGCalSD.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (kpedro88) [reply SimG4CMS/Calo/interface/HGCalSD.h]: In this case, the comment seems wrong (it mentions the pre-TDR version)

File: SimG4CMS/Calo/src/HGCalNumberingScheme.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (kpedro88) [reply SimG4CMS/Calo/src/HGCalNumberingScheme.cc]: Right, I'm wondering why it's always 1. (I guess it never comes into the function with a useful value?)

**Comment by davidlange6** [comment]
please test

**Comment by bsunanda** [comment]
@ianna @kpedro88 @civanch Can you approve this PR?

**Comment by ianna** [comment]
@bsunanda - please, check if cornerAll is needed here https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23129/27838/llvm-analysis/report-a77242.html#EndPath

File: SimG4CMS/Calo/interface/HGCalSD.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (kpedro88) [reply SimG4CMS/Calo/interface/HGCalSD.h]: Can you clarify the comment in the code?

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: SimG4CMS/Calo/interface/HGCalSD.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+#include "SimG4CMS/Calo/interface/CaloSD.h"
+#include "SimG4Core/Notification/interface/BeginOfJob.h"
+#include "SimG4CMS/Calo/interface/HGCalNumberingScheme.h"
+#include "SimG4CMS/Calo/interface/HGCMouseBite.h"
+#include "DetectorDescription/Core/interface/DDsvalues.h"
+
+#include "G4String.hh"
+#include <string>
+#include <TH1F.h>
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply SimG4CMS/Calo/interface/HGCalSD.h]: I am removing them

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
@bsunanda please fix the comment in HGCalSD.h that refers to pre-TDR

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
+1

**Comment by bsunanda** [comment]
@ianna @civanch Please approve this PR

**Comment by bsunanda** [comment]
@ianna @civanch Please approve this PR

**Comment by ianna** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
@kpedro88 @civanch I see several differences in the D19 scenario, that elsewhere is claimed to be frozen. So what should be considered really to be kept here?

**Comment by bsunanda** [comment]
The new simulation takes totally new path as mentioned by Kevin. I am not sure what could cause the fluctuation in SimHits for 2023 workflows. There are places where division by 10.0 is replaced by multiplication by 0.1. But this cannot affect things in a big way except the earlier bug fix which had spurious hits and corrected for.

**Comment by civanch** [comment]
@bsunanda , @kpedro88 , are you sure that geometry is not changed down to the last digit?

**Comment by bsunanda** [comment]
I did not change any of the xml file nor the algorithms for the old geometry

**Comment by fabiocos** [comment]
+1

**Pull request merged**